library/src/main/java/com/novoda/downloadmanager/lib/DatabaseHelper.java

```diff
@@ -0,0 +1,231 @@
+     * Note: to support downgrades, creating a table should always drop it first if it already
+     * exists.
+     */
+    @Override
+    public void onUpgrade(final SQLiteDatabase db, int oldV, final int newV) {
+        if (oldV == 31) {
+            // 31 and 100 are identical, just in different codelines. Upgrading from 31 is the
+            // same as upgrading from 100.
+            oldV = 100;
+        } else if (oldV < 100) {
+            // no logic to upgrade from these older version, just recreate the DB
+            Log.i("Upgrading downloads database from version " + oldV + " to version " + newV + ", which will destroy all old data");
+            oldV = 99;
+        } else if (oldV > newV) {
+            // user must have downgraded software; we have no way to know how to downgrade the
+            // DB, so just recreate it
+            Log.i("Downgrading downloads database from version " + oldV + " (current version is " + newV + "), destroying all old data");
+            oldV = 99;
+        }
+
+        for (int version = oldV + 1; version <= newV; version++) {
+            upgradeTo(db, version);
+        }
+    }
+
+    /**
+     * Upgrade database from (version - 1) to version.
+     */
+    private void upgradeTo(SQLiteDatabase db, int version) {
+        switch (version) {
+            case 100:
+                createDownloadsTable(db);
+                break;
+
+            case 101:
+                createHeadersTable(db);
+                break;
+
+            case 102:
+                addColumn(db, tableName, Downloads.Impl.COLUMN_ALLOW_ROAMING, "INTEGER NOT NULL DEFAULT 0");
+                addColumn(db, tableName, Downloads.Impl.COLUMN_ALLOWED_NETWORK_TYPES, "INTEGER NOT NULL DEFAULT 0");
+                break;
+
+            case 103:
+                addColumn(db, tableName, Downloads.Impl.COLUMN_IS_VISIBLE_IN_DOWNLOADS_UI, "INTEGER NOT NULL DEFAULT 1");
+                makeCacheDownloadsInvisible(db);
+                break;
+
+            case 104:
+                addColumn(db, tableName, Downloads.Impl.COLUMN_BYPASS_RECOMMENDED_SIZE_LIMIT, "INTEGER NOT NULL DEFAULT 0");
+                break;
+
+            case 105:
+                fillNullValues(db);
+                break;
+
+            case 106:
+                addColumn(db, tableName, Downloads.Impl.COLUMN_MEDIAPROVIDER_URI, "TEXT");
+                addColumn(db, tableName, Downloads.Impl.COLUMN_DELETED, "BOOLEAN NOT NULL DEFAULT 0");
+                break;
+
+            case 107:
+                addColumn(db, tableName, Downloads.Impl.COLUMN_ERROR_MSG, "TEXT");
+                break;
+
+            case 108:
+                addColumn(db, tableName, Downloads.Impl.COLUMN_ALLOW_METERED, "INTEGER NOT NULL DEFAULT 1");
+                break;
+
+            case 109:
+                addColumn(db, tableName, Downloads.Impl.COLUMN_BIG_PICTURE, "TEXT");
+                break;
+            case 110:
+                createBatchesTable(db);
```

Comment:
so you do do some new stuff in this file - I would delete all the other cases and start from here